llvm/test/Transforms/Attributor/do-not-replace-addrspacecast-with-constantpointernull.ll

```diff
@@ -0,0 +1,34 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -S -passes=attributor %s -o - | FileCheck %s
+
+define i32 @bar(ptr %p0, ptr addrspace(5) %p5) {
+; CHECK-LABEL: define i32 @bar(
+; CHECK-SAME: ptr nofree readonly captures(none) [[P0:%.*]], ptr addrspace(5) nofree readonly [[P5:%.*]]) #[[ATTR0:[0-9]+]] {
+; CHECK-NEXT:    [[ICMP:%.*]] = icmp eq ptr addrspace(5) [[P5]], addrspacecast (ptr null to ptr addrspace(5))
+; CHECK-NEXT:    [[SELECT:%.*]] = select i1 [[ICMP]], ptr [[P0]], ptr null
+; CHECK-NEXT:    [[LOAD:%.*]] = load i32, ptr [[SELECT]], align 4
+; CHECK-NEXT:    ret i32 [[LOAD]]
+;
+  %icmp = icmp eq ptr addrspace(5) %p5, addrspacecast (ptr null to ptr addrspace(5))
+  %select = select i1 %icmp, ptr %p0, ptr null
+  %load = load i32, ptr %select, align 4
+  ret i32 %load
+}
+
+define i32 @foo(ptr %p0, ptr %p1, <2 x ptr addrspace(5)> %ptrvec) {
+; CHECK-LABEL: define i32 @foo(
+; CHECK-SAME: ptr nofree readonly captures(none) [[P0:%.*]], ptr nofree noundef nonnull readonly align 16 captures(none) dereferenceable(16) [[P1:%.*]], <2 x ptr addrspace(5)> [[PTRVEC:%.*]]) #[[ATTR0]] {
+; CHECK-NEXT:    [[LOADVEC:%.*]] = load <2 x ptr addrspace(5)>, ptr [[P1]], align 16
+; CHECK-NEXT:    [[ICMPVEC:%.*]] = icmp eq <2 x ptr addrspace(5)> [[LOADVEC]], <ptr addrspace(5) addrspacecast (ptr null to ptr addrspace(5)), ptr addrspace(5) addrspacecast (ptr null to ptr addrspace(5))>
+; CHECK-NEXT:    [[ICMP:%.*]] = extractelement <2 x i1> [[ICMPVEC]], i32 1
+; CHECK-NEXT:    [[SELECT:%.*]] = select i1 [[ICMP]], ptr [[P0]], ptr null
+; CHECK-NEXT:    [[LOAD:%.*]] = load i32, ptr [[SELECT]], align 4
+; CHECK-NEXT:    ret i32 [[LOAD]]
+;
+  %loadvec = load <2 x ptr addrspace(5)>, ptr %p1, align 16
+  %icmpvec = icmp eq <2 x ptr addrspace(5)> %loadvec, <ptr addrspace(5) addrspacecast (ptr null to ptr addrspace(5)), ptr addrspace(5) addrspacecast (ptr null to ptr addrspace(5))>
+  %icmp = extractelement <2 x i1> %icmpvec, i32 1
+  %select = select i1 %icmp, ptr %p0, ptr null
+  %load = load i32, ptr %select, align 4
+  ret i32 %load
+}
```

Comment:
Also add the versions with addrspace(1) instead of 5 which will use 0 init

Reply:
But you should, that would make the test more useful

Reply:
Added, though honestly I don't think it matters at this point. We don't currently have a way to notify the middle end whether a `nullptr` in a given address space is zero or not, and there's no middle-end hook for that either, so it won't actually change anything in practice. I'd be surprised if it made any difference right now.

Reply:
But it will catch the change in the future patch. There's no plus to not making this a target test that will expose the difference